File: frontend/src/types/forum.ts
export type ForumComment = {
  commentId: number;
  commentUuid: string;
  postId: number;
  postUuid: string;
  courseId: number;
  courseUuid: string;
  authorUserId: number;
  authorUserUuid: string;
  authorEmail: string;
  authorName: string;
  rootCommentId: number | null;
  rootCommentUuid: string | null;
  replyToCommentId: number | null;
  replyToCommentUuid: string | null;
  replyToAuthorName: string | null;
  content: string;
  commentKind: string;
  metadataJson: Record<string, unknown> | null;
  isDeleted: boolean;
  deletedAt: string | null;
  replyCount: number;
  createdAt: string;
  updatedAt: string;
};

export type ForumPost = {
  postId: number;
  postUuid: string;
  courseId: number;
  courseUuid: string;
  authorUserId: number;
  authorUserUuid: string;
  authorEmail: string;
  authorName: string;
  title: string | null;
  content: string;
  postKind: string;
  metadataJson: Record<string, unknown> | null;
  isPinned: boolean;
  pinnedAt: string | null;
  commentCount: number;
  previewComments: ForumComment[];
  createdAt: string;
  updatedAt: string;
};

export type PaginatedForumPostResponse = {
  items: ForumPost[];
  page: number;
  pageSize: number;
  total: number;
  totalPages: number;
};

export type PaginatedForumCommentResponse = {
  items: ForumComment[];
  page: number;
  pageSize: number;
  total: number;
  totalPages: number;
};

export type CreateForumPostPayload = {
  title?: string;
  content: string;
  postKind?: string;
  metadataJson?: Record<string, unknown> | null;
};

export type UpdateForumPostPayload = {
  title?: string | null;
  content?: string;
  metadataJson?: Record<string, unknown> | null;
};

export type CreateForumCommentPayload = {
  content: string;
  commentKind?: string;
  replyToCommentUuid?: string | null;
  metadataJson?: Record<string, unknown> | null;
};

export type UpdateForumCommentPayload = {
  content?: string;
  metadataJson?: Record<string, unknown> | null;
};

File: services/communication-service/app/schemas/forum.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class CourseForumCommentCreateRequest(BaseModel):
    content: str = Field(..., min_length=1)
    commentKind: str = Field(default="user", min_length=1, max_length=20)
    replyToCommentUuid: str | None = None
    metadataJson: dict[str, Any] | None = None


class CourseForumCommentUpdateRequest(BaseModel):
    content: str | None = Field(default=None, min_length=1)
    metadataJson: dict[str, Any] | None = None


class CourseForumCommentRead(BaseModel):
    commentId: int
    commentUuid: str
    postId: int
    postUuid: str
    courseId: int
    courseUuid: str
    authorUserId: int
    authorUserUuid: str
    authorEmail: str
    authorName: str
    rootCommentId: int | None
    rootCommentUuid: str | None
    replyToCommentId: int | None
    replyToCommentUuid: str | None
    replyToAuthorName: str | None
    content: str
    commentKind: str
    metadataJson: dict[str, Any] | None
    isDeleted: bool
    deletedAt: datetime | None
    replyCount: int
    createdAt: datetime
    updatedAt: datetime


class PaginatedCourseForumCommentResponse(BaseModel):
    items: list[CourseForumCommentRead]
    page: int
    pageSize: int
    total: int
    totalPages: int


class CourseForumPostCreateRequest(BaseModel):
    title: str | None = Field(default=None, max_length=200)
    content: str = Field(..., min_length=1)
    postKind: str = Field(default="user", min_length=1, max_length=20)
    metadataJson: dict[str, Any] | None = None


class CourseForumPostUpdateRequest(BaseModel):
    title: str | None = Field(default=None, max_length=200)
    content: str | None = Field(default=None, min_length=1)
    metadataJson: dict[str, Any] | None = None


class CourseForumPostRead(BaseModel):
    postId: int
    postUuid: str
    courseId: int
    courseUuid: str
    authorUserId: int
    authorUserUuid: str
    authorEmail: str
    authorName: str
    title: str | None
    content: str
    postKind: str
    metadataJson: dict[str, Any] | None
    isPinned: bool
    pinnedAt: datetime | None
    commentCount: int
    previewComments: list[CourseForumCommentRead]
    createdAt: datetime
    updatedAt: datetime


class PaginatedCourseForumPostResponse(BaseModel):
    items: list[CourseForumPostRead]
    page: int
    pageSize: int
    total: int
    totalPages: int
